proxy.go

```diff
@@ -162,6 +162,84 @@ func (rp *reverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	requestDuration.With(s.labels).Observe(since)
 }
 
+type ResponseWriterWithCode interface {
+	http.ResponseWriter
+	StatusCode() int
+}
+
+func runReq(ctx context.Context, s *scope, startTime time.Time, retryNum int, rp func(http.ResponseWriter, *http.Request), rw http.ResponseWriter, srw *statResponseWriter, req *http.Request, err error) error {
+	// test if we could do the rerun
+	numr := len(s.host.replica.cluster.replicas)
+	if numr > 1 && retryNum <= numr {
+		for i := 0; i <= retryNum; i++ {
+			rp(rw, req)
+			err = ctx.Err()
+			if err == nil { //nolint: gocritic
+				// The request has been successfully proxied.
+				//	 since := time.Since(startTime).Seconds()
+				// proxiedResponseDuration.With(s.labels).Observe(since)
+
+				// cache.FSResponseWriter pushes status code to srw on Finalize/Fail actions
+				// but they didn't happen yet, so manually propagate the status code from crw to srw.
+				if crw, ok := rw.(ResponseWriterWithCode); ok {
+					srw.statusCode = crw.StatusCode()
+				}
+
+				fmt.Println(srw.statusCode)
+
+				// StatusBadGateway response is returned by http.ReverseProxy when
+				// it cannot establish connection to remote host.
+				if srw.statusCode == http.StatusBadGateway {
+					if i == retryNum {
+						s.host.penalize()
+						q := getQuerySnippet(req)
+						err = fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
+						respondWith(srw, err, srw.statusCode)
```

Comment:
There is a bug here, the srw Write method cannot write the statusCode. I am still checking the issue.

Reply:
Solved